compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java

```diff
@@ -8185,4 +8166,295 @@ private boolean isEndOfTypeList(SyntaxKind nextTokenKind) {
+        return parseTableConstructorOrQuery(peek().kind, tableKeyword, keySpecifier);
+    }
+
+    private STNode parseTableConstructorOrQuery(SyntaxKind nextTokenKind, STNode tableKeyword, STNode keySpecifier) {
+        switch (nextTokenKind) {
+            case OPEN_BRACKET_TOKEN:
+                return parseTableConstructorExprRhs(tableKeyword, keySpecifier);
+            case KEY_KEYWORD:
+                keySpecifier = parseKeySpecifier();
+                return parseTableConstructorOrQueryRhs(peek().kind, tableKeyword, keySpecifier);
+            default:
+                Solution solution = recover(peek(), ParserRuleContext.TABLE_KEYWORD_RHS, tableKeyword, keySpecifier);
+
+                // If the parser recovered by inserting a token, then try to re-parse the same
+                // rule with the inserted token. This is done to pick the correct branch
+                // to continue the parsing.
+                if (solution.action == Action.REMOVE) {
+                    return solution.recoveredNode;
+                }
+
+                return parseTableConstructorOrQuery(solution.tokenKind, tableKeyword, keySpecifier);
+        }
+    }
+
+    private STNode parseTableConstructorOrQueryRhs(STNode tableKeyword, STNode keySpecifier) {
+        return parseTableConstructorOrQueryRhs(peek().kind, tableKeyword, keySpecifier);
+    }
+
+    private STNode parseTableConstructorOrQueryRhs(SyntaxKind nextTokenKind, STNode tableKeyword, STNode keySpecifier) {
+        switch (nextTokenKind) {
+            case FROM_KEYWORD:
+                return parseQueryExprRhs(parseQueryConstructType(tableKeyword, keySpecifier));
+            case OPEN_BRACKET_TOKEN:
+                return parseTableConstructorExprRhs(tableKeyword, keySpecifier);
+            default:
+                Solution solution = recover(peek(), ParserRuleContext.TABLE_CONSTRUCTOR_OR_QUERY_RHS,
+                    tableKeyword, keySpecifier);
+
+                // If the parser recovered by inserting a token, then try to re-parse the same
+                // rule with the inserted token. This is done to pick the correct branch
+                // to continue the parsing.
+                if (solution.action == Action.REMOVE) {
+                    return solution.recoveredNode;
+                }
+
+                return parseTableConstructorOrQueryRhs(solution.tokenKind, tableKeyword, keySpecifier);
+        }
+    }
+
+    /**
+     * Parse query expression.
+     * <code>query-expr-rhs := query-pipeline select-clause</code>
+     *
+     * @param queryConstructType queryConstructType that precedes this rhs
+     * @return Parsed node
+     */
+    private STNode parseQueryExprRhs(STNode queryConstructType) {
+        switchContext(ParserRuleContext.QUERY_EXPRESSION);
+        STNode queryPipeline = parseQueryPipeline();
+        STNode selectClause = parseSelectClause();;
+        endContext();
```

Comment:
We can see if there are more clauses after the select clause as well, and give an error. 

Can do the same thing as in `parseQueryPipeline`, but simply log and error.

Reply:
So this logic also go inside parseQueryExprRhs if we are to go ahead with one of above mentioned approaches?